components/util/memory.rs

```diff
@@ -338,22 +338,187 @@ impl MemoryProfiler {
         // each reporter once we have enough of them.
         //
         // If anything goes wrong with a reporter, we just skip it.
+        let mut forest = ReportsForest::new();
         for reporter in self.reporters.values() {
             let (chan, port) = channel();
             if reporter.collect_reports(MemoryReportsChan(chan)) {
                 if let Ok(reports) = port.recv() {
-                    for report in reports {
-                        let mebi = 1024f64 * 1024f64;
-                        println!("{:12.2}: {}", (report.size as f64) / mebi, report.name);
+                    for report in reports.iter() {
+                        forest.insert(&report.path, report.size);
                     }
                 }
             }
         }
+        forest.print();
 
+        println!("|");
+        println!("End memory reports");
         println!("");
     }
 }
 
+/// A collection of one or more reports with the same initial path segment. A ReportsTree
+/// containing a single node is described as "degenerate".
+struct ReportsTree {
+    /// For leaf nodes, this is the sum of the sizes of all reports that mapped to this location.
+    /// For interior nodes, this is the sum of the sizes of all its child nodes.
+    size: u64,
+
+    /// For leaf nodes, this is the count of all reports that mapped to this location.
+    /// For interor nodes, this is always zero.
+    count: u32,
+
+    /// The segment from the report path that maps to this node.
+    path_seg: String,
+
+    /// Child nodes.
+    children: Vec<ReportsTree>,
+}
+
+impl ReportsTree {
+    fn new(path_seg: String) -> ReportsTree {
+        ReportsTree {
+            size: 0,
+            count: 0,
+            path_seg: path_seg,
+            children: vec![]
+        }
+    }
+
+    // Searches the tree's children for a path_seg match, and returns the index if there is a
+    // match.
+    fn find_child(&self, path_seg: &String) -> Option<usize> {
+        let mut i = 0;
+        for child in self.children.iter() {
+            if child.path_seg == *path_seg {
+                return Some(i);
+            }
+            i += 1;
+        }
+        None
+    }
+
+    // Insert the path and size into the tree, adding any nodes as necessary.
+    fn insert(&mut self, path: &[String], size: u64) {
+        // Modifying a recursive data structure while traversing it is difficult in safe Rust. So
+        // this whole function uses raw pointers.
+        unsafe {
+            let mut t: *mut ReportsTree = self;
+            for path_seg in path.iter() {
+                let i = match (*t).find_child(&path_seg) {
+                    Some(i) => i,
+                    None => {
+                        let new_t = ReportsTree::new(path_seg.clone());
+                        (*t).children.push(new_t);
+                        (*t).children.len() - 1
+                    },
+                };
+                t = &mut (*t).children[i];
```

Comment:
I believe this should be possible in safe code, by explicitly moving out of `t` before reassigning, e.g.:

```rust
let tmp = t;
t = &mut tmp.children[i];
```

cf.

- http://stackoverflow.com/questions/27083544/trouble-porting-c-code-to-rust
- http://www.reddit.com/r/rust/comments/2zm0kh/how_to_mutably_traverse_a_linked_list_efficiently/